ostd/src/mm/vm_space.rs

```diff
@@ -91,47 +118,42 @@ impl VmSpace {
         Ok(self.pt.cursor_mut(va).map(|pt_cursor| {
             let activation_lock = self.activation_lock.read();
 
-            let mut activated_cpus = CpuSet::new_empty();
-
-            for cpu in all_cpus() {
-                // The activation lock is held; other CPUs cannot activate this `VmSpace`.
-                let ptr =
-                    ACTIVATED_VM_SPACE.get_on_cpu(cpu).load(Ordering::Relaxed) as *const VmSpace;
-                if ptr == self as *const VmSpace {
-                    activated_cpus.add(cpu);
-                }
-            }
-
             CursorMut {
                 pt_cursor,
                 activation_lock,
-                flusher: TlbFlusher::new(activated_cpus, disable_preempt()),
+                flusher: TlbFlusher::new(self.cpus.load(), disable_preempt()),
             }
         })?)
     }
 
     /// Activates the page table on the current CPU.
     pub(crate) fn activate(self: &Arc<Self>) {
         let preempt_guard = disable_preempt();
+        let cpu = preempt_guard.current_cpu();
 
-        // Ensure no mutable cursors (which holds read locks) are alive.
+        let last_ptr = ACTIVATED_VM_SPACE.load();
+
+        if last_ptr == Arc::as_ptr(self) {
+            return;
+        }
+
+        // Ensure no mutable cursors (which holds read locks) are alive before
+        // we add the CPU to the CPU set.
```

Comment:
This PR looks good to me! I only have a concern which I overlooked before, not introduced by this PR. 

I am a bit concerned because the current usage of `CursorMut` does not always drop immediately after use; there are also scenarios like holding it for a longer period in `handle_page_fault_around` when processing a range of page faults consecutively. In such cases, I feel it might lead to prolonged spinning here, which could impact performance. 

I'm not entirely sure, but my understanding is that this `activation_lock` is intended to address the TLB flusher and the vmspace `clear` situation. So, could we reference this lock in the TLB flusher and add the lock only during the do remote flush operation, and obtain a `CursorMut` during `clear` to do the same lock effect? It's somewhat like turning a big lock into a smaller one.

Reply:
TLB flusher lives as long as the cursor lives. If the VM Space is activated on another CPU when a cursor on this CPU is unmapping something, the TLB flusher should flush the newcomer CPU's TLB. But it isn't possible if activation is not synchronized.

A fine grained approach to identify non-unmapping cursors is the only solution to the problem you have. It is added to the TODO list.